routing/cross_mwm_index_graph.cpp

```diff
@@ -106,54 +172,105 @@ void CrossMwmIndexGraph::GetTwins(Segment const & s, bool isOutgoing, std::vecto
+    vector<ms::LatLon> const & latLons =
+        isOutgoing ? GetLatLon(it->second.m_outgoing, s) : GetLatLon(it->second.m_ingoing, s);
     for (string const & name : neighboringMwm)
     {
-      auto const addTransitionSegments = [&](NumMwmId numMwmId, TRoutingMappingPtr const & mapping)
-      {
-        if (isOutgoing)
-        {
-          mapping->m_crossContext.ForEachIngoingNodeNearPoint(latLon, [&](IngoingCrossNode const & node){
-            AddTransitionSegment(mapping->m_segMapping, node.m_nodeId, numMwmId, twins);
-          });
-        }
-        else
+      NumMwmId const numMwmId = m_numMwmIds->GetId(CountryFile(name));
+      auto const addFirstValidSegment = [&](TRoutingMappingPtr const & mapping) {
+        for (auto const & ll : latLons)
         {
-          mapping->m_crossContext.ForEachOutgoingNodeNearPoint(latLon, [&](OutgoingCrossNode const & node){
-            AddTransitionSegment(mapping->m_segMapping, node.m_nodeId, numMwmId, twins);
-          });
+          if (isOutgoing)
+          {
+            mapping->m_crossContext.ForEachIngoingNodeNearPoint(
+                ll, [&](IngoingCrossNode const & node) {
+                  AddFirstValidSegment(mapping->m_segMapping, numMwmId, node.m_nodeId, twins);
+                });
+          }
+          else
+          {
+            mapping->m_crossContext.ForEachOutgoingNodeNearPoint(
+                ll, [&](OutgoingCrossNode const & node) {
+                  AddFirstValidSegment(mapping->m_segMapping, numMwmId, node.m_nodeId, twins);
+                });
+          }
         }
       };
 
-      if (!LoadWith(m_numMwmIds->GetId(CountryFile(name)), addTransitionSegments))
+      if (!LoadWith(numMwmId, addFirstValidSegment))
         continue;  // mwm was not loaded.
     }
   };
 
   LoadWith(s.GetMwmId(), getTwins);
   my::SortUnique(twins);
+
+  for (Segment const & t : twins)
+    CHECK_NOT_EQUAL(s.GetMwmId(), t.GetMwmId(), ());
 }
 
-void CrossMwmIndexGraph::GetEdgeList(Segment const & /* s */,
-                                     bool /* isOutgoing */, std::vector<SegmentEdge> & /* edges */) const
+void CrossMwmIndexGraph::GetEdgeList(Segment const & s, bool isOutgoing,
+                                     vector<SegmentEdge> & edges)
 {
   // @TODO(bykoianko) It's necessary to check if mwm of |s| contains an A* cross mwm section
   // and if so to use it. If not, osrm cross mwm sections should be used.
-  NOTIMPLEMENTED();
+
+  CHECK(IsTransition(s, !isOutgoing), ("The segment is not a transition segment. IsTransition(",
+                                       s, ",", !isOutgoing, ") returns false."));
+
+  // Note. According to cross-mwm OSRM sections there're node id which could be ingoing and outgoing
+  // at the same time. For example in Berlin mwm on Nordlicher Berliner Ring (A10) near crossing
+  // with A11
+  // there's such node id. It's an extremely rare case. There're probably several such node id
```

Comment:
It's not an easy task you know. To do that I need to take Node Id (I know one), then to look up feature id(s) which covers the node id, and then to look up (in a binary file) osm id which corresponds to the node id. Taking into account that we're going to remove osrm support in the feature I'd prefer not to find the osm id(s).